src/Microsoft.Azure.SignalR.Common/ServiceCommunicationScope.cs

```diff
@@ -0,0 +1,48 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+
+namespace Microsoft.Azure.SignalR.Common
+{
+    /// <summary>
+    /// Represents a generic nestable scope for grouping operations with the same service communication policy.
+    /// Derived classes must ensure that the service protocol can understand and transmit this policy to the service.
+    /// </summary>
+    /// The following is ensured for all send operations to the service within the scope:
+    /// -  the same service connection is used in ServiceTransportType.Persistent mode
+    /// -  the same service connection is used for calls outside of hub call context
+    /// -  only the value of the most nested scope will flow along with the messages
+    /// 
+    /// <typeparam name="TNestableProps">Type representing policy to flow</typeparam>
+    public class ServiceCommunicationScope<TNestableProps> : IDisposable
```

Comment:
is it `public` intentionally and is designed to be used by our customers?

Reply:
Customer's won't be able to use it as its only constructor is protected. Its derived class is public so the base has to be public too, e.g. 
~~~
public class TraceScope : ServiceCommunicationScope<bool> 
{}
~~~

Reply:
Customer can write his own derived class? is it expected?

Reply:
Customers are only expected to use derived class (such as `TraceScope' in example above).
I don't see any good reason why customers would want to derive themselves.

Reply:
moving nestable scopes to a separate PR